src/zcash/circuit/prfs.tcc

```diff
@@ -0,0 +1,139 @@
+template<typename FieldT>
+class PRF_gadget : gadget<FieldT> {
+private:
+    std::shared_ptr<block_variable<FieldT>> block;
+    std::shared_ptr<sha256_compression_function_gadget<FieldT>> hasher;
+    std::shared_ptr<digest_variable<FieldT>> result;
+
+public:
+    PRF_gadget(
+        protoboard<FieldT>& pb,
+        pb_variable<FieldT>& ZERO,
+        bool a,
+        bool b,
+        bool c,
+        bool d,
+        pb_variable_array<FieldT> x,
+        boost::optional<pb_variable_array<FieldT>> y,
+        std::shared_ptr<digest_variable<FieldT>> result
+    ) : gadget<FieldT>(pb), result(result) {
+
+        pb_linear_combination_array<FieldT> IV = SHA256_default_IV(pb);
+
+        pb_variable_array<FieldT> discriminants;
+        discriminants.emplace_back(a ? ONE : ZERO);
+        discriminants.emplace_back(b ? ONE : ZERO);
+        discriminants.emplace_back(c ? ONE : ZERO);
+        discriminants.emplace_back(d ? ONE : ZERO);
+
+        if (!y) {
+            // Create y and pad it with zeroes.
+            y = pb_variable_array<FieldT>();
+            while (y->size() < 256) {
+                y->emplace_back(ZERO);
+            }
+        }
+
+        block.reset(new block_variable<FieldT>(pb, {
+            discriminants,
+            x,
+            *y
```

Comment:
I guess I find it odd that `y` is optional rather than explicitly passing zeroes.